typescript/ts-lowerings/src/org/jetbrains/dukat/tsLowerings/IntroduceMissingConstructors.kt

```diff
@@ -1,45 +1,74 @@
+        when {
+            classDeclaration.typeParameters.isEmpty() -> constructors
+            else -> {
+                val typesMapping = classDeclaration.typeParameters.withIndex().map { (index, type) ->
+                    val resolvedType = actualTypes.getOrNull(index) ?: type.defaultValue
+                    type.name to resolvedType!!
+                }.toMap()
+                val resolver = ResolveTypeParamsInConstructor(typesMapping)
+                constructors.map { resolver.lowerConstructorDeclaration(it, null) }
+            }
         }
+
+
+    override fun lowerClassDeclaration(
+        declaration: ClassDeclaration,
+        owner: NodeOwner<ModuleDeclaration>?
+    ): TopLevelDeclaration? {
+        val classWithConstructors = parentConstructorsMap[declaration.uid]
+        val declarationResolved = classWithConstructors.addConstructorsInto(declaration)
         return super.lowerClassDeclaration(declarationResolved, owner)
     }
 
     override fun lower(source: SourceSetDeclaration): SourceSetDeclaration {
         declarationResolver = TopLevelDeclarationResolver(source)
 
-        val classDeclarations = declarationResolver.asIterable().filterIsInstance(ClassDeclaration::class.java).toposort { classDeclaration ->
-            classDeclaration.parentEntities.mapNotNull { declarationResolver.resolveRecursive(it.typeReference?.uid) as? ClassDeclaration }
-        }
+        val classDeclarations =
+            declarationResolver.asIterable().filterIsInstance<ClassDeclaration>().toposort { classDeclaration ->
+                classDeclaration.parentEntities.mapNotNull { declarationResolver.resolveRecursive(it.typeReference?.uid) as? ClassDeclaration }
+            }
 
         classDeclarations.forEach { classDeclaration ->
-            val constructors = classDeclaration.members.filterIsInstance(ConstructorDeclaration::class.java)
-            val constructorsResolved = if (constructors.isEmpty()) {
-                classDeclaration.parentEntities.flatMap { parentEntity ->
-                    parentEntity.typeReference?.uid?.let { parentConstructorsMap[it] } ?: emptyList()
+            val constructors = classDeclaration.members.filterIsInstance<ConstructorDeclaration>()
+            val constructorsResolved = when {
+                constructors.isNotEmpty() -> constructors
+                else -> classDeclaration.parentEntities.flatMap { parentEntity ->
+                    val uid = parentEntity.typeReference?.uid
+                    val parentConstructors = uid?.let { parentConstructorsMap[it] }
+                    parentConstructors?.getConstructorsWithResolvedTypeParams(parentEntity.typeArguments) ?: emptyList()
                 }
-            } else {
-                constructors
             }
 
-            parentConstructorsMap[classDeclaration.uid] = constructorsResolved
+            parentConstructorsMap[classDeclaration.uid] = ClassWithConstructors(classDeclaration, constructorsResolved)
```

Comment:
Seems like it can't, because we get the information about the class by `uid` of `HeritageClauseDeclaration` instance here and have no the `ClassDeclaration` instance for the `uid` inside the code snippet:
https://github.com/Kotlin/dukat/pull/451/files#diff-d8a9dd29c23dbca67ebd858db4a988df9a106025bfde2342f36a650775b2bd6bR66

Reply:
Hm, okay, I am not sure, but if it is unique id, it should be instance of `ClassDeclaration` too, should't it?
Or it can be instance of other class but with the same `uid`?
I expected, that `typeReference` with the same `uid` is de-facto `ClassDeclaration`
But it is not so important

Reply:
Unfortunately, `typeReference` doesn't contain any kind of declaration.
It's a nullable instance of `ReferenceDeclaration` class with few properties:

<img width="1055" alt="Screenshot 2021-09-02 at 13 19 56" src="https://user-images.githubusercontent.com/10776615/131827406-589c7b6f-dd7a-4feb-81e2-5350a24e7ecb.png">

Reply:
Okay, I see, thanks. So yes, you are right. We need leave `uid` as a key